<a href="https://colab.research.google.com/github/nasira-banu/nasirabanussprojectiiith/blob/main/Copy_of_AIML_III_Module_01_Lab_02_Machine_Learning_terms_and_metrics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Machine Learning terms and metrics

Module 1, Lab 2

In this lab, we will show a part of the ML pipeline by using the California Housing dataset. There are 20640 samples, each with 8 attributes like income of the block, age of the houses per district etc. The task is to predict the cost of the houses per district. We will use the scikit-learn library to load the data and perform some basic data preprocessing and model training. We will also show how to evaluate the model using some common metrics, split the data into training and testing sets, and use cross-validation to get a better estimate of the model's performance.

In [8]:
import numpy as np
from sklearn import datasets
import matplotlib.pyplot as plt

rng = np.random.default_rng(seed=42)

In [69]:
dataset = datasets.fetch_california_housing()
# Dataset description
print("Description of the attributes", dataset.DESCR)
print("Size of the data set" ,dataset.data.shape) #number of rows and colums
print("Data = ", dataset.data) #data


Description of the attributes .. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

    :Number of Instances: 20640

    :Number of Attributes: 8 numeric, predictive attributes and the target

    :Attribute Information:
        - MedInc        median income in block group
        - HouseAge      median house age in block group
        - AveRooms      average number of rooms per household
        - AveBedrms     average number of bedrooms per household
        - Population    block group population
        - AveOccup      average number of household members
        - Latitude      block group latitude
        - Longitude     block group longitude

    :Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,0

Given below are the list of target values. These correspond to the house value derived considering all the 8 input features and are continuous values. We should use regression models to predict these values but we will start with a simple classification model for the sake of simplicity. We need to just round off the values to the nearest integer and use a classification model to predict the house value.

In [10]:
print("Orignal target values:", dataset.target)

dataset.target = dataset.target.astype(float)

print("Target values after conversion:", dataset.target)
print("Input variables shape:", dataset.data.shape)
print("Output variables shape:", dataset.target.shape)

Orignal target values: [4.526 3.585 3.521 ... 0.923 0.847 0.894]
Target values after conversion: [4.526 3.585 3.521 ... 0.923 0.847 0.894]
Input variables shape: (20640, 8)
Output variables shape: (20640,)


The simplest model to use for classification is the K-Nearest Neighbors model. We will use this model to predict the house value with a K value of 1. We will also use the accuracy metric to evaluate the model.

In [11]:
def NN1(traindata, trainlabel, query):
    """
    This function takes in the training data, training labels and a query point
    and returns the predicted label for the query point using the nearest neighbour algorithm

    traindata: numpy array of shape (n,d) where n is the number of samples and d is the number of features
    trainlabel: numpy array of shape (n,) where n is the number of samples
    query: numpy array of shape (d,) where d is the number of features

    returns: the predicted label for the query point which is the label of the training data which is closest to the query point
    """
    diff = (traindata - query)  # find the difference between features. Numpy automatically takes care of the size here
    sq = diff **2 # square the differences
    dist = sq.sum(1)  # add up the squares
    label = trainlabel[np.argmin(dist)]
    return label


def NN(traindata, trainlabel, testdata):
    """
    This function takes in the training data, training labels and test data
    and returns the predicted labels for the test data using the nearest neighbour algorithm

    traindata: numpy array of shape (n,d) where n is the number of samples and d is the number of features
    trainlabel: numpy array of shape (n,) where n is the number of samples
    testdata: numpy array of shape (m,d) where m is the number of test samples and d is the number of features

    returns: the predicted labels for the test data which is the label of the training data which is closest to each test point
    """
    predlabel = np.array([NN1(traindata, trainlabel, i) for i in testdata])
    return predlabel

We will also define a 'random classifier', which randomly allots labels to each sample

In [12]:
def RandomClassifier(traindata, trainlabel, testdata):
    """
    This function takes in the training data, training labels and test data
    and returns the predicted labels for the test data using the random classifier algorithm

    In reality, we don't need these arguments but we are passing them to keep the function signature consistent with other classifiers

    traindata: numpy array of shape (n,d) where n is the number of samples and d is the number of features
    trainlabel: numpy array of shape (n,) where n is the number of samples
    testdata: numpy array of shape (m,d) where m is the number of test samples and d is the number of features

    returns: the predicted labels for the test data which is a random label from the training data
    """

    classes = np.unique(trainlabel)
    rints = rng.integers(low=0, high=len(classes), size=len(testdata))
    predlabel = classes[rints]
    return predlabel

We need a metric to evaluate the performance of the model. Let us define a metric 'Accuracy' to see how good our learning algorithm is. Accuracy is the ratio of the number of correctly classified samples to the total number of samples. The higher the accuracy, the better the algorithm. We will use the accuracy metric to evaluate and compate the performance of the K-Nearest Neighbors model and the random classifier.

In [13]:
def Accuracy(gtlabel, predlabel):
    """
    This function takes in the ground-truth labels and predicted labels
    and returns the accuracy of the classifier

    gtlabel: numpy array of shape (n,) where n is the number of samples
    predlabel: numpy array of shape (n,) where n is the number of samples

    returns: the accuracy of the classifier which is the number of correct predictions divided by the total number of predictions
    """
    assert len(gtlabel) == len(
        predlabel
    ), "Length of the ground-truth labels and predicted labels should be the same"
    correct = (
        gtlabel == predlabel
    ).sum()  # count the number of times the groundtruth label is equal to the predicted label.
    return correct / len(gtlabel)

Let us make a function to split the dataset with the desired probability. We will use this function to split the dataset into training and testing sets. We will use the training set to train the model and the testing set to evaluate the model.

In [14]:
def split(data, label, percent):
    # generate a random number for each sample
    rnd = rng.random(len(label))
    split1 = rnd < percent
    split2 = rnd >= percent

    split1data = data[split1, :]
    split1label = label[split1]
    split2data = data[split2, :]
    split2label = label[split2]
    return split1data, split1label, split2data, split2label

We will reserve 20% of our dataset as the test set. We will not change this portion throughout our experiments

In [15]:
testdata, testlabel, alltraindata, alltrainlabel = split(
    dataset.data, dataset.target, 20 / 100
)
print("Number of test samples:", len(testlabel))
print("Number of train samples:", len(alltrainlabel))
print("Percent of test data:", len(testlabel) * 100 / len(dataset.target), "%")

Number of test samples: 4144
Number of train samples: 16496
Percent of test data: 20.07751937984496 %


## Experiments with splits

Let us reserve some of our train data as a validation set

In [16]:
traindata, trainlabel, valdata, vallabel = split(
    alltraindata, alltrainlabel, 75 / 100)

What is the accuracy of our classifiers on the train dataset?

In [17]:
trainpred = NN(traindata, trainlabel, traindata)
trainAccuracy = Accuracy(trainlabel, trainpred)
print("Training accuracy using nearest neighbour algorithm:", trainAccuracy*100, "%")

trainpred = RandomClassifier(traindata, trainlabel, traindata)
trainAccuracy = Accuracy(trainlabel, trainpred)
print("Training accuracy using random classifier: ", trainAccuracy*100, "%")

Training accuracy using nearest neighbour algorithm: 100.0 %
Training accuracy using random classifier:  0.03234152652005175 %


For nearest neighbour, the train accuracy is always 1. The accuracy of the random classifier is close to 1/(number of classes) which is 0.1666 in our case. This is because the random classifier randomly assigns a label to each sample and the probability of assigning the correct label is 1/(number of classes). Let us predict the labels for our validation set and get the accuracy. This accuracy is a good estimate of the accuracy of our model on unseen data.

In [18]:
valpred = NN(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using nearest neighbour algorithm:", valAccuracy*100, "%")


valpred = RandomClassifier(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using random classifier:", valAccuracy*100, "%")

Validation accuracy using nearest neighbour algorithm: 0.9205426356589148 %
Validation accuracy using random classifier: 0.0 %


Validation accuracy of nearest neighbour is considerably less than its train accuracy while the validation accuracy of random classifier is the same. However, the validation accuracy of nearest neighbour is twice that of the random classifier. Now let us try another random split and check the validation accuracy. We will see that the validation accuracy changes with the split. This is because the validation set is small and the accuracy is highly dependent on the samples in the validation set. We can get a better estimate of the accuracy by using cross-validation.

In [19]:
traindata, trainlabel, valdata, vallabel = split(
    alltraindata, alltrainlabel, 75 / 100)
valpred = NN(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using nearest neighbour algorithm:", valAccuracy*100, "%")

Validation accuracy using nearest neighbour algorithm: 0.9748964172556666 %


You can run the above cell multiple times to try with different random splits.
We notice that the accuracy is different for each run, but close together.

Now let us compare it with the accuracy we get on the test dataset.

In [20]:
testpred = NN(alltraindata, alltrainlabel, testdata)
testAccuracy = Accuracy(testlabel, testpred)

print("Test accuracy:", testAccuracy*100, "%")

Test accuracy: 0.9893822393822393 %


### Try it out for yourself and answer:
1. How is the accuracy of the validation set affected if we increase the percentage of validation set? What happens when we reduce it?
2. How does the size of the train and validation set affect how well we can predict the accuracy on the test set using the validation set?
3. What do you think is a good percentage to reserve for the validation set so that thest two factors are balanced?

Answer for both nearest neighbour and random classifier. You can note down the values for your experiments and plot a graph using  <a href=https://matplotlib.org/stable/gallery/lines_bars_and_markers/step_demo.html#sphx-glr-gallery-lines-bars-and-markers-step-demo-py>plt.plot<href>. Check also for extreme values for splits, like 99.9% or 0.1%

> Exercise: Try to implement a 3 nearest neighbour classifier and compare the accuracy of the 1 nearest neighbour classifier and the 3 nearest neighbour classifier on the test dataset. You can use the KNeighborsClassifier class from the scikit-learn library to implement the K-Nearest Neighbors model. You can set the number of neighbors using the n_neighbors parameter. You can also use the accuracy_score function from the scikit-learn library to calculate the accuracy of the model.

In [25]:
#Question and Exercise
traindata, trainlabel, valdata, vallabel = split(
    alltraindata, alltrainlabel, 85 / 100)

trainpred = NN(traindata, trainlabel, traindata)
trainAccuracy = Accuracy(trainlabel, trainpred)
print("Training accuracy using nearest neighbour algorithm:", trainAccuracy*100, "%")

trainpred = RandomClassifier(traindata, trainlabel, traindata)
trainAccuracy = Accuracy(trainlabel, trainpred)
print("Training accuracy using random classifier: ", trainAccuracy*100, "%")

valpred = NN(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using nearest neighbour algorithm:", valAccuracy*100, "%")


valpred = RandomClassifier(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using random classifier:", valAccuracy*100, "%")

testpred = NN(alltraindata, alltrainlabel, testdata)
testAccuracy = Accuracy(testlabel, testpred)

print("Test accuracy NN:", testAccuracy*100, "%")

testpred = RandomClassifier(alltraindata, alltrainlabel, testdata)
testAccuracy = Accuracy(testlabel, testpred)

print("Test accuracy Random Classifier:", testAccuracy*100, "%")



Training accuracy using nearest neighbour algorithm: 100.0 %
Training accuracy using random classifier:  0.042668183757644716 %
Validation accuracy using nearest neighbour algorithm: 0.8627773212818405 %
Validation accuracy using random classifier: 0.12325390304026293 %
Test accuracy NN: 0.9893822393822393 %
Test accuracy Random Classifier: 0.0 %


In [26]:
#Questions and Exercise
traindata, trainlabel, valdata, vallabel = split(
    alltraindata, alltrainlabel, 99.9 / 100)

trainpred = NN(traindata, trainlabel, traindata)
trainAccuracy = Accuracy(trainlabel, trainpred)
print("Training accuracy using nearest neighbour algorithm:", trainAccuracy*100, "%")

trainpred = RandomClassifier(traindata, trainlabel, traindata)
trainAccuracy = Accuracy(trainlabel, trainpred)
print("Training accuracy using random classifier: ", trainAccuracy*100, "%")

valpred = NN(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using nearest neighbour algorithm:", valAccuracy*100, "%")


valpred = RandomClassifier(traindata, trainlabel, valdata)
valAccuracy = Accuracy(vallabel, valpred)
print("Validation accuracy using random classifier:", valAccuracy*100, "%")

testpred = NN(alltraindata, alltrainlabel, testdata)
testAccuracy = Accuracy(testlabel, testpred)

print("Test accuracy NN:", testAccuracy*100, "%")

testpred = RandomClassifier(alltraindata, alltrainlabel, testdata)
testAccuracy = Accuracy(testlabel, testpred)

print("Test accuracy Random Classifier:", testAccuracy*100, "%")



Training accuracy using nearest neighbour algorithm: 100.0 %
Training accuracy using random classifier:  0.024273317555676924 %
Validation accuracy using nearest neighbour algorithm: 0.0 %
Validation accuracy using random classifier: 0.0 %
Test accuracy NN: 0.9893822393822393 %
Test accuracy Random Classifier: 0.07239382239382239 %


In [31]:
#Excercise

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder # Import LabelEncoder

x = dataset.data
y = dataset.target

# Convert continuous target values to discrete labels if necessary
label_encoder = LabelEncoder() # Initialize LabelEncoder
y = label_encoder.fit_transform(y) # Transform y to discrete labels

# Splititting the dataset into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# 1-NN classifier
knn_1 = KNeighborsClassifier(n_neighbors=1)
knn_1.fit(x_train, y_train)
y_pred_1 = knn_1.predict(x_test)
accuracy_1 = accuracy_score(y_test, y_pred_1)
print(f"Accuracy of 1-NN classifier: {accuracy_1}")

# 3-NN classifier
knn_3 = KNeighborsClassifier(n_neighbors=3)
knn_3.fit(x_train, y_train)
y_pred_3 = knn_3.predict(x_test)
accuracy_3 = accuracy_score(y_test, y_pred_3)
print(f"Accuracy of 3-NN classifier: {accuracy_3}")

Accuracy of 1-NN classifier: 0.008963178294573644
Accuracy of 3-NN classifier: 0.005813953488372093


For the given data set the accuracy of 1-NN classifier is higher than a 3-NN one

## Multiple Splits

One way to get more accurate estimates for the test accuracy is by using <b>cross-validation</b>. Here, we will try a simple version, where we do multiple train/val splits and take the average of validation accuracies as the test accuracy estimation. Here is a function for doing this. Note that this function will take a long time to execute. You can reduce the number of splits to make it faster.

In [32]:
def AverageAccuracy(alldata, alllabel, splitpercent, iterations, classifier=NN):
    """
    This function takes in the data, labels, split percentage, number of iterations and classifier function
    and returns the average accuracy of the classifier

    alldata: numpy array of shape (n,d) where n is the number of samples and d is the number of features
    alllabel: numpy array of shape (n,) where n is the number of samples
    splitpercent: float which is the percentage of data to be used for training
    iterations: int which is the number of iterations to run the classifier
    classifier: function which is the classifier function to be used

    returns: the average accuracy of the classifier
    """
    accuracy = 0
    for ii in range(iterations):
        traindata, trainlabel, valdata, vallabel = split(
            alldata, alllabel, splitpercent
        )
        valpred = classifier(traindata, trainlabel, valdata)
        accuracy += Accuracy(vallabel, valpred)
    return accuracy / iterations  # average of all accuracies

In [23]:
avg_acc = AverageAccuracy(alltraindata, alltrainlabel, 75 / 100, 10, classifier=NN)
print("Average validation accuracy:", avg_acc*100, "%")
testpred = NN(alltraindata, alltrainlabel, testdata)

print("Test accuracy:", Accuracy(testlabel, testpred)*100, "%")

Average validation accuracy: 0.8653748009396562 %
Test accuracy: 0.9893822393822393 %


In [36]:
#Increasing the number of iterations
avg_acc = AverageAccuracy(alltraindata, alltrainlabel, 75 / 100, 25, classifier=NN)
print("Average validation accuracy for 25 iterations :", avg_acc*100, "%")
testpred = NN(alltraindata, alltrainlabel, testdata)

print("Test accuracy for 25 iterations:", Accuracy(testlabel, testpred)*100, "%")


avg_acc = AverageAccuracy(alltraindata, alltrainlabel, 75 / 100, 50, classifier=NN)
print("Average validation accuracy for 50 iterations:", avg_acc*100, "%")
testpred = NN(alltraindata, alltrainlabel, testdata)

print("Test accuracy for 50 iterations :", Accuracy(testlabel, testpred)*100, "%")

Average validation accuracy for 25 iterations : 0.8518650777230318 %
Test accuracy for 25 iterations: 0.9893822393822393 %
Average validation accuracy for 59 iterations: 0.8850896157942325 %
Test accuracy for 50 iterations : 0.9893822393822393 %


In [38]:
#For 99.9% split
avg_acc = AverageAccuracy(alltraindata, alltrainlabel, 99.9 / 100, 25, classifier=NN)
print("Average validation accuracy for 25 iterations :", avg_acc*100, "%")
testpred = NN(alltraindata, alltrainlabel, testdata)

print("Test accuracy for 25 iterations:", Accuracy(testlabel, testpred)*100, "%")


avg_acc = AverageAccuracy(alltraindata, alltrainlabel, 99.9 / 100, 50, classifier=NN)
print("Average validation accuracy for 50 iterations:", avg_acc*100, "%")
testpred = NN(alltraindata, alltrainlabel, testdata)

print("Test accuracy for 50 iterations :", Accuracy(testlabel, testpred)*100, "%")

Average validation accuracy for 25 iterations : 0.7478260869565218 %
Test accuracy for 25 iterations: 0.9893822393822393 %
Average validation accuracy for 50 iterations: 0.6214932126696833 %
Test accuracy for 50 iterations : 0.9893822393822393 %


This is a very simple way of doing cross-validation. There are many well-known algorithms for cross-validation, like k-fold cross-validation, leave-one-out etc. This will be covered in detail in a later module. For more information about cross-validation, check <a href=https://en.wikipedia.org/wiki/Cross-validation_(statistics)>Cross-validatioin (Wikipedia)</a>

### Questions
1. Does averaging the validation accuracy across multiple splits give more consistent results? *Yes, as observed with 10, 25 and 50 overall it does give more consistent results.*

2. Does it give more accurate estimate of test accuracy? *Yes, it gives more accurate estimate (Validation accuracy using nearest neighbour algorithm without averaging : 0.8627773212818405 %, Average validation accuracy for 50 iterations: 0.8850896157942325 %)*


3. What is the effect of the number of iterations on the estimate? Do we get a better estimate with higher iterations? *As observed with 25 iterations over 10 iterations calculated above we see that the average accuracy has decreased a little bit but with 50 iterations it has increased sufficiently. So substantially high number of iterations might give a better estimate*

4. Consider the results you got for the previous questions. Can we deal with a very small train dataset or validation dataset by increasing the iterations? *The validation accuracy has substantially increased from 0.0% to .747% and .621 % for 25 and 50 iterations respectively.*


> Exercise: How does the accuracy of the 3 nearest neighbour classifier change with the number of splits? How is it affected by the split size? Compare the results with the 1 nearest neighbour classifier.
The average validation and test accuracy has increased significantly wrt 1-NN Classifier
1-NN : Average validation accuracy: 0.8653748009396562 %
Test accuracy: 0.9893822393822393 %
3-NN : Average validation accuracy for 3-NN: 33.636723609236086 %
Test accuracy for 3-NN: 1.0174418604651163 %
The results have been same even with increasing splits and split size. *italicized text*

In [55]:
#Increasing the number of iterations
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)
knn_1 = KNeighborsClassifier(n_neighbors=1)
knn_1.fit(x_train, y_train)

knn_3 = KNeighborsClassifier(n_neighbors=3)
knn_3.fit(x_train, y_train)

def AverageAccuracy3(alldata, alllabel, splitpercent, iterations, classifier=knn_3):

    accuracy = 0
    for ii in range(iterations):
        traindata, trainlabel, valdata, vallabel = split(
            alldata, alllabel, splitpercent
        )
        # Use the predict method of the classifier
        valpred = classifier.predict(valdata)
        accuracy += Accuracy(vallabel, valpred)
    return accuracy / iterations  # average of all accuracies

avg_acc_3nn = AverageAccuracy3(x_train, y_train, 0.75, 10, classifier=knn_3)
print("Average validation accuracy for 3-NN:", avg_acc_3nn * 100, "%")
testpred = NN(x_train, y_train, x_test)
print("Test accuracy for 3-NN:", Accuracy(y_test, testpred) * 100, "%")

avg_acc_3nn = AverageAccuracy3(x_train, y_train, 0.80, 25, classifier=knn_3)
print("Average validation accuracy for 3-NN:", avg_acc_3nn * 100, "%")
testpred = NN(x_train, y_train, x_test)
print("Test accuracy for 3-NN:", Accuracy(y_test, testpred) * 100, "%")



Average validation accuracy for 3-NN: 33.636723609236086 %
Test accuracy for 3-NN: 1.0174418604651163 %
Average validation accuracy for 3-NN: 33.71963158058233 %
Test accuracy for 3-NN: 1.0174418604651163 %
